### DISCUSSION TITLE
Terragrunt Imports Roadmap

### DISCUSSION THREAD

**Discussion by bakavets** [gruntwork-io/terragrunt#3517] [General]
Hi there! Hope you are doing well! 

Could you share any information about [imports](https://terragrunt.gruntwork.io/docs/rfc/imports/)?

Ref: [Implement imports RFC](https://github.com/gruntwork-io/terragrunt/issues/1566)

Is there any ETA for that feature?

Thanks for your reply in advance!

**Comment by josh-padnick** [comment]
Hello! FYI, the best person to answer this is @yhakbar, however he's OOO until next week. We'll post back then!

**Comment by yhakbar** [comment]
Hey @bakavets ,

I wasn't here when the initial imports RFC was proposed, but I've read through it.

To be honest, I'm not sure we should invest time in developing this feature, or that it's necessary. The fact that Terragrunt configurations can't be merged through chained configuration includes right now results in configurations that are easier to read.

e.g.

In the scenario where you have the following `import`s:
A <-- B <-- C
A' <-- B <-- C

Where C depends on B, and B depends on both A and A', you have four different configurations you have to read to know what's going on in C, and two of them aren't even referenced in the C file. This feels to me like the kind of abstraction that makes Terragrunt harder to use, and it doesn't provide any functional advantage, as far as I can tell. If someone was to instead structure their configurations like this:

B <-- C
A <-- C
A' <-- C

You still have to read four configurations, but you can now read C alone and know everywhere that configurations _could_ be relevant to C. This is especially important in CI systems, as you might want to trigger changes for C when anything it `include`s changes. This is pretty straightforward right now, but would be a lot more complex in a world with chained `import`s. 

In addition, it's also easy to reason about how precedence is determined. You can read the file top to bottom, and see where configurations will get overwritten (unless you decide to put includes after `inputs` for some reason). In a world with chained `import`s, you have two dimensions to track: both the depth of the `import`, and the position of the `import` configuration block in HCL in a given file.

Also, I definitely don't think these should be called `import`s, as that conflicts with OpenTofu `import` blocks, and that will likely confuse people.

Personally, I'm for closing out the RFC as will not do, and achieving DRY-er configurations via [Stacks](https://github.com/gruntwork-io/terragrunt/issues/3313).

I'll share my thoughts on the linked issue as well, and see if the community agrees.

**Reply by tdaniely-ludeo** [reply]
> Hey @bakavets ,
> 
> I wasn't here when the initial imports RFC was proposed, but I've read through it.
> 
> To be honest, I'm not sure we should invest time in developing this feature, or that it's necessary. The fact that Terragrunt configurations can't be merged through chained configuration includes right now results in configurations that are easier to read.
> 
> e.g.
> 
> In the scenario where you have the following `import`s: A <-- B <-- C A' <-- B <-- C
> 
> Where C depends on B, and B depends on both A and A', you have four different configurations you have to read to know what's going on in C, and two of them aren't even referenced in the C file. This feels to me like the kind of abstraction that makes Terragrunt harder to use, and it doesn't provide any functional advantage, as far as I can tell. If someone was to instead structure their configurations like this:
> 
> B <-- C A <-- C A' <-- C
> 
> You still have to read four configurations, but you can now read C alone and know everywhere that configurations _could_ be relevant to C. This is especially important in CI systems, as you might want to trigger changes for C when anything it `include`s changes. This is pretty straightforward right now, but would be a lot more complex in a world with chained `import`s.
> 
> In addition, it's also easy to reason about how precedence is determined. You can read the file top to bottom, and see where configurations will get overwritten (unless you decide to put includes after `inputs` for some reason). In a world with chained `import`s, you have two dimensions to track: both the depth of the `import`, and the position of the `import` configuration block in HCL in a given file.
> 
> Also, I definitely don't think these should be called `import`s, as that conflicts with OpenTofu `import` blocks, and that will likely confuse people.
> 
> Personally, I'm for closing out the RFC as will not do, and achieving DRY-er configurations via [Stacks](https://github.com/gruntwork-io/terragrunt/issues/3313).
> 
> I'll share my thoughts on the linked issue as well, and see if the community agrees.

The suggestion that multiple includes, or multi-layer inclusion is bad because we loose view of the structure, seems shaky at best. It's disregarding that many people would like to express structure in other ways, such as folder layout.

A layout like the following, or a variation of it, is ideal for many many users.

repo
> root.hcl
> cloud
>> cloud.hcl
>> region
>>> region.hcl
>>> app
>>>> app.hcl
>>>> unit
>>>>> terragrunt.hcl

And in this case we're having to rewrite the structure again in each terragrunt.hcl which is often hundreds or more for a large organization. When we've already defined the structure we want explicitly by structuring the configuration it self. And it is very clear to us what goes where and why. And additionally we end up having this long variable names and 20 lines that repeat themselves in every file.

I'm sorry if this comes out aggressive, but there's a lot of frustration with the direction terragrunt is taking on these points. It is highly opinionated, and seemingly the opinion is based on lofty design fixations which are not based in every day reality.

Maybe this is vim / emacs culture fighting desktop culture, but this whole discussion makes very little sense, to a lot of people.

The tool definitely needs to enable and not constrain. Configuration design should come from compositional patterns, not tooling design-by-constraints.

I hope i managed to express my opinion and frustration without offending anyone. 🤷🏼‍♂️
And thank you very much to all contributors. ❤️
